packages/@react-aria/interactions/test/useInteractOutside.test.js

```diff
@@ -442,3 +442,150 @@ describe('useInteractOutside (iframes)', function () {
+    const onInteractOutside = jest.fn();
+    const {shadowRoot, cleanup} = createShadowRootAndRender(
+      <App onInteractOutside={onInteractOutside} />
+    );
+
+    const popover = shadowRoot.getElementById('popover');
+    fireEvent.mouseDown(popover);
+    fireEvent.mouseUp(popover);
+
+    expect(onInteractOutside).not.toHaveBeenCalled();
+    cleanup();
+  });
+
+  it('triggers when clicking outside the popover', function () {
+    const onInteractOutside = jest.fn();
+    const  {cleanup} = createShadowRootAndRender(
+      <App onInteractOutside={onInteractOutside} />
+    );
+
+    // Clicking on the document body outside the shadow DOM
+    fireEvent.mouseDown(document.body);
+    fireEvent.mouseUp(document.body);
+
+    expect(onInteractOutside).toHaveBeenCalledTimes(1);
+    cleanup();
+  });
+
+  it('triggers when clicking a button outside the shadow dom altogether', function () {
+    const onInteractOutside = jest.fn();
+    const {cleanup} = createShadowRootAndRender(
+      <App onInteractOutside={onInteractOutside} />
+    );
+    // Button outside shadow DOM and component
+    const button = document.createElement('button');
+    document.body.appendChild(button);
+
+    fireEvent.mouseDown(button);
+    fireEvent.mouseUp(button);
+
+    expect(onInteractOutside).toHaveBeenCalledTimes(1);
+    document.body.removeChild(button);
+    cleanup();
+  });
+});
+
+describe('useInteractOutside shadow DOM extended tests', function () {
+  // Setup function similar to previous tests, but includes a dynamic element scenario
+  function createShadowRootAndRender(ui) {
+    const shadowHost = document.createElement('div');
+    document.body.appendChild(shadowHost);
+    const shadowRoot = shadowHost.attachShadow({mode: 'open'});
+
+    function WrapperComponent() {
+      return ReactDOM.createPortal(ui, shadowRoot);
+    }
+
+    render(<WrapperComponent />);
+    return {shadowRoot, cleanup: () => document.body.removeChild(shadowHost)};
+  }
+
+  function App({onInteractOutside, includeDynamicElement = false}) {
+    const ref = useRef(null);
+    useInteractOutside({ref, onInteractOutside});
+
+    useEffect(() => {
+      if (includeDynamicElement) {
+        const dynamicEl = document.createElement('div');
+        dynamicEl.id = 'dynamic-outside';
+        document.body.appendChild(dynamicEl);
+
+        return () => document.body.removeChild(dynamicEl);
+      }
+    }, [includeDynamicElement]);
+
+    return (
+      <div>
+        <div id="outside-popover" />
+        <div id="popover" ref={ref}>
+          <div id="inside-popover" />
+        </div>
+      </div>
+    );
+  }
+
+  it('correctly identifies interaction with dynamically added external elements', function () {
+    const onInteractOutside = jest.fn();
+    const {cleanup} = createShadowRootAndRender(
+      <App onInteractOutside={onInteractOutside} includeDynamicElement />
+    );
+
+    // Wait for dynamic element to be added
+    setTimeout(() => {
```

Comment:
useFakeTimers, then run them just after this setTimeout, that way there's no danger of the assertions running after the test has completed

Reply:
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

Reply:
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a)